transaction.go

```diff
@@ -90,21 +90,14 @@ type Transaction struct {
 
 // checksum calculates the checksum of a transaction excluding the checksum
 // field of each page
-func (t Transaction) checksum() ([crypto.HashSize]byte, error) {
-	page := t.firstPage
-
-	// Marshall all the pages and calculate the checksum
-	var data []byte
-	for page != nil {
-		bytes, err := page.Marshal()
-		if err != nil {
-			return [crypto.HashSize]byte{}, build.ExtendErr("Failed to marshall page", err)
-		}
-		// Ignore the transactionChecksum field (first 32 bytes)
-		data = append(data, bytes[crypto.HashSize:]...)
-		page = page.nextPage
+func (t Transaction) checksum() (c checksum) {
```

Comment:
It's worth noting for the sake of future PRs we might want to consider, the `checksum()` function necessarily marshals everything that you are saving to disk. Then after we compute the checksum we marshal it again. So maybe we can keep the data around and then pass it to the thing writing to disk.

It's not completely trivial because your pages could be scattered all over  the file, so you'd still be writing to disk 1 page at a time. And additionally, there's not really a performance hit because our marshaling is basically just `binary.PutUint64` and `append`. But still wanted to highlight it since the `wal` is supposed to be extremely high performance.

Reply:
Sounds fun!